File: functions/src/sportsday/temporary_authentication.ts
import * as functions from 'firebase-functions'
import * as admin from 'firebase-admin'
import {v4 as uuid} from 'uuid'

export const sportsDayTemporaryAuth = functions.region('europe-west2').https.onCall(() => {
    const id = 'sd-' + uuid()
    return admin.auth().createCustomToken(id, { isSportsDayUser: true })
})

File: lib/data/sportsday/temporary_authentication.dart
import 'package:cloud_functions/cloud_functions.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:mymgs/data/local_database.dart';
import 'package:sembast/sembast.dart';

final _functions = FirebaseFunctions.instanceFor(region: 'europe-west2');
final _auth = FirebaseAuth.instance;
final store = StoreRef<String, bool>('sports-day');
Future<void> authenticateSportsDay() async {
  final response = await _functions.httpsCallable('sportsDayTemporaryAuth')();
  final db = await getDb();
  await store.record('temporary-auth').put(db, true);

  await _auth.signInWithCustomToken(response.data);
}

Future<void> resetSportsDayAuth() async {
  final db = await getDb();
  await store.record('temporary-auth').delete(db);
}

// returns true if the user is currently signed in using a temporary Sports Day session
Future<bool> isSportsDayAuth() async {
  final db = await getDb();
  return await store.record('temporary-auth').get(db) ?? false;
}
